# Part 1. Equation of a Slime

How many late days are you using for this assignment? None

In [187]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold, cross_val_score
from sklearn.linear_model import LinearRegression  

## 1. Loading the dataset

In [188]:
# Using pandas load the dataset
df = pd.read_csv('science_data_large.csv')  
# Output the first 15 rows of the data
print(df.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(df.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [189]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df.drop(columns=['Size nm^3'])  
y = df['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [190]:
# Use sklearn to train a model on the training set
model = LinearRegression().fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_point = pd.DataFrame([[469,647]], columns=X.columns)

prediction = model.predict(sample_point)
print("Predicted Output:", prediction)

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print("Model Score:", score)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)
equation = f"h(x) = {intercept:.4f} + " + " + ".join([f"{coef:.4f}x_{i}" for i, coef in enumerate(coefficients, start=1)])
print(equation)

Predicted Output: [664984.89630433]
Model Score: 0.8552472077276096
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.4795834081
h(x) = -409391.4796 + 866.1464x_1 + 1032.6951x_2


Write the linear equation of a slime: (example equation: $E = mc^2$)
$$
y = -409391.4796 + 866.1464 * Temp + 1032.6951 * Mols
$$

Given that we only consider 5 significant figures, the updated equation is 
$$
y = -409390 + 866.15 * Temp + 1032.7 * Mols
$$



Report on score and explain meaning:

The default score function property of SKLearn model for linear regression returns the coefficient of determination $R^2$, best possible score is 1.0 indicating a perfect fit and worst is -infinity meaning the model is performing worse than just predicting the mean. A score of 0 represnts a constant model that always predicts the expected value of y without taking into factor the input features.

The model score of 0.85524 means that the linear regression model explains about 85.52% of the variance in the target variable. This suggests that the model captures some patterns in the data but still leaves about 14.48% of the variance unexplained. The model does a good job at capturing the underlying relationship between the features and the target variable, but there is still about 14.48% of the variance that the model cannot explain.


## 4. Use Cross Validation

In [191]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kf = KFold(n_splits=5, shuffle=True, random_state=42)
cross_score = cross_val_score(model, X, y, cv=kf)

# Report on their finding and their significance
print("Cross-val R² scores:", cross_score)
print(f"Mean R² Score: {cross_score.mean():.4f}")
print(f"Standard Deviation: {cross_score.std():.4f}")

Cross-val R² scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean R² Score: 0.8597
Standard Deviation: 0.0184


Write findings here:


- The Mean $R^2$ score represents that the model explains 85.97% of the variance in the target variable on average across the 5 cross-validation folds. Therefore, it is a good fit to data, but can be better.
- The standard deviation of 0.0184 shows good variability in the model’s performance. The relatively low standard deviation suggests that the model’s performance is constant over several subsets of data 

## 5. Using Polynomial Regression

In [192]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import PolynomialFeatures

# Create polynomial regression model (degree 2)
poly_model = make_pipeline(PolynomialFeatures(degree=2), LinearRegression())
poly_model.fit(X_train, y_train)

# Cross-validation for polynomial regression
kf = KFold(n_splits=5, shuffle=True, random_state=42)
poly_cv_scores = cross_val_score(poly_model, X, y, cv=kf)
print("Polynomial Regression CV Scores:", poly_cv_scores)
print("Mean Polynomial CV Score:", poly_cv_scores.mean())
print("Standard Deviation:", poly_cv_scores.std())

score_poly = poly_model.score(X_test, y_test)
print("Model Score:", score_poly)

coeff = poly_model.named_steps['linearregression'].coef_
inter = poly_model.named_steps['linearregression'].intercept_

print("Intercept:", inter)
print("Coefficients:", coeff)

Polynomial Regression CV Scores: [1. 1. 1. 1. 1.]
Mean Polynomial CV Score: 1.0
Standard Deviation: 0.0
Model Score: 1.0
Intercept: 2.047908492386341e-05
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27198206e-07  1.26473309e-11
  2.00000000e+00  2.85714287e-02]


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$$
y = 2.047908492386341 \times 10^{-5} + (12 \cdot \text{Temp}) + (-1.27198206 \times 10^{-7} \cdot \text{Mols}) + (1.26473309 \times 10^{-11} \cdot \text{Temp}^2) +
(2 \cdot \text{Mols}^2) + (0.0285714287 \cdot \text{Temp} \cdot \text{Mols})
$$

Given that we only consider 5 significant figures, the updated equation is 

$$
y=(12⋅Temp)+(2⋅Mols^2)+(0.028571⋅Temp⋅Mols)
$$

Report on the score and interpret:


The score function returns the coefficient of determination $R^2$ as explained in Part 3. 1.0 indicating a perfect fit of the model and data. The model explains 100% of the variance in the target variable. This means the model's predictions match the actual observed values exactly.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [193]:
# Load the dataset. Then train and evaluate the classification models.
df = pd.read_csv('ckd_feature_subset.csv')
X = df.drop(columns=['Target_ckd'])  
y = df['Target_ckd']

from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "Support Vector Machines": SVC(),
    "k-Nearest Neighbors": KNeighborsClassifier(),
    "Neural Networks": MLPClassifier(max_iter=1000)  
}

results = {}

for model_name, model in models.items():
    kf = KFold(n_splits=5, shuffle=True, random_state=42)
    cross_score = cross_val_score(model, X, y, cv=kf)
    results[model_name] = {
        "Mean Accuracy": np.mean(cross_score),
        "Standard Deviation": np.std(cross_score)
    }

results_df = pd.DataFrame(results).T
print(results_df)

                         Mean Accuracy  Standard Deviation
Logistic Regression           0.856559            0.066269
Support Vector Machines       0.928172            0.047601
k-Nearest Neighbors           0.927957            0.052440
Neural Networks               0.941505            0.047178


/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments
The table above summarizes the mean accuracy and standard deviation of the four classification models evaluated using 5-fold cross-validation on the CKD dataset.
- The Neural Network model achieved the highest mean accuracy of 0.9351 with a standard deviation of 0.0405. This implies that the model performed vbetter than other models and also exhibited high consistency across different folds of cross-validation. Here's some reasons why:
- Neural networks have the ability to learn complex patterns and interactions in the data through multiple layers of interconnected nodes. This allows identification of more intricate patterns.  
- Unlike Logistic Regression, which assumes a linear relationship, neural networks can learn feature interactions without explicit manual engineering.



In [194]:
# Experiments with Neural Network.
nn_configs = {
    "Config 1 (1 layer, 32 neurons, relu, adam)": MLPClassifier(hidden_layer_sizes=(32,), activation='relu', solver='adam', max_iter=1000, random_state=42),
    "Config 2 (2 layers, 64-32 neurons, tanh, sgd)": MLPClassifier(hidden_layer_sizes=(64, 32), activation='tanh', solver='sgd', learning_rate_init=0.01, max_iter=1000, random_state=42),
    "Config 3 (3 layers, 128-64-32 neurons, relu, adam)": MLPClassifier(hidden_layer_sizes=(128, 64, 32), activation='relu', solver='adam', max_iter=1000, random_state=42),
}

nn_results = {}
kf = KFold(n_splits=5, shuffle=True, random_state=42)

for config_name, model in nn_configs.items():
    cross_score = cross_val_score(model, X, y, cv=kf)
    nn_results[config_name] = {
        "Mean Accuracy": np.mean(cross_score),
        "Standard Deviation": np.std(cross_score)
    }

nn_results_df = pd.DataFrame(nn_results).T
print(nn_results_df)


                                                    Mean Accuracy  \
Config 1 (1 layer, 32 neurons, relu, adam)               0.928387   
Config 2 (2 layers, 64-32 neurons, tanh, sgd)            0.941505   
Config 3 (3 layers, 128-64-32 neurons, relu, adam)       0.967527   

                                                    Standard Deviation  
Config 1 (1 layer, 32 neurons, relu, adam)                    0.047183  
Config 2 (2 layers, 64-32 neurons, tanh, sgd)                 0.031258  
Config 3 (3 layers, 128-64-32 neurons, relu, adam)            0.035340  


## Results and Conclusion for Neural Network Experiments
- Config 3 outperformed the other configurations with a mean accuracy of 0.9675 and a relatively low standard deviation of 0.0353. This indicates high performance and consistency across different folds of cross-validation.
- The substantial improvement from Config 2 to Config 3 (from 0.9415 to 0.9675) indicates that increasing the number of layers and neurons significantly enhances the model’s ability to learn complex relationships. The three-layer configuration with 128, 64, and 32 neurons allowed for deeper representation learning with a further increase in mean accuracy. 
- The choice of the Adam optimizer in Config 1 and Config 3 helped those models converge faster and achieve better performance than traditional methods like stochastic gradient descent (SGD) used in Config 2.


The experimentation illustrates that:
Increasing Model Capacity: Adding more layers and neurons allows the model to capture richer representations of the data, significantly improving classification performance.
Tuning Activation Functions and Optimizers: The selection of appropriate activation functions and optimizers can substantially affect training efficiency and performance. In this case, using ReLU and Adam optimized configurations yielded the best results.

